aspnetcore/fundamentals/servers/kestrel/diagnostics.md

```diff
@@ -0,0 +1,132 @@
+
+| Logging Category Name | Logging Events |
+|--|--|
+| `Microsoft.AspNetCore.Server.Kestrel` |  ApplicationError, ConnectionHeadResponseBodyWrite, ApplicationNeverCompleted, RequestBodyStart, RequestBodyDone, RequestBodyNotEntirelyRead, RequestBodyDrainTimedOut, ResponseMinimumDataRateNotSatisfied, InvalidResponseHeaderRemoved, HeartbeatSlow |
+| `Microsoft.AspNetCore.Server.Kestrel.BadRequests` | ConnectionBadRequest, RequestProcessingError, RequestBodyMinimumDataRateNotSatisfied |
+| `Microsoft.AspNetCore.Server.Kestrel.Connections` | ConnectionAccepted, ConnectionStart, ConnectionStop, ConnectionPause, ConnectionResume, ConnectionKeepAlive, ConnectionRejected, ConnectionDisconnect, NotAllConnectionsClosedGracefully, NotAllConnectionsAborted, ApplicationAbortedConnection |
+| `Microsoft.AspNetCore.Server.Kestrel.Http2` | Http2ConnectionError, Http2ConnectionClosing, Http2ConnectionClosed, Http2StreamError, Http2StreamResetAbort, HPackDecodingError, HPackEncodingError, Http2FrameReceived, Http2FrameSending, Http2MaxConcurrentStreamsReached |
+| `Microsoft.AspNetCore.Server.Kestrel.Http3` | Http3ConnectionError, Http3ConnectionClosing, Http3ConnectionClosed, Http3StreamAbort, Http3FrameReceived, Http3FrameSending |
+
+### Connection logging
+
+// TODO: WIP
+
+```csharp
+using System.Diagnostics;
+using System.Net;
+var builder = WebApplication.CreateBuilder(args);
+builder.WebHost.ConfigureLogging(logging =>
+    logging.AddFilter((category, level) =>
+        category.Equals("Microsoft.AspNetCore.Server.Kestrel.Core.Internal.LoggingConnectionMiddleware") && level >= LogLevel.Trace));
+builder.WebHost.ConfigureKestrel(o =>
+{
+    o.Listen(IPAddress.Loopback, 5000, listenOptions => listenOptions.UseConnectionLogging());
+    o.Listen(IPAddress.Loopback, 5001, listenOptions =>
+    {
+        listenOptions.UseHttps();
+        // For logging decrypted traffic
+        listenOptions.UseConnectionLogging();
+    });
+});
+var app = builder.Build();
+app.MapGet("/", () => "Hello world");
+app.Run();
+```
+
+## Metrics
+
+Metrics is a representation of data measures over intervals of time, for example, requests per second. Metrics data allows observation of the state of an app at a high-level. Kestrel metrics are emitted using `EventCounter`.
+
+> Unfortunately, the `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. Unlike, as implied by them the name, they do not always contain the number of new connections or TLS handshakes per second, but rather the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`. It is **recommended** that consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
```

Comment:
```suggestion
> [!NOTE]
> The `connections-per-second` and `tls-handshakes-per-second` counters are named incorrectly. The counters:
>  * Do ***not*** always contain the number of new connections or TLS handshakes per second
>  * Display the number of new connection or TLS handshakes in the last update interval as requested as the consumer of Events via the `EventCounterIntervalSec` argument in the `filterPayload` to `KestrelEventSource`.
>
> We **recommend** consumers of these counters scale the metric value based on the `DisplayRateTimeScale` of one second.
```

Reply:
Oh wow! That's a **LOT** easier to understand